ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/ThreadNet/Network.hs

```diff
@@ -165,6 +171,22 @@ plainTestNodeInitialization pInfo = TestNodeInitialization
     , tniProtocolInfo = pInfo
     }
 
+-- | Compute the chain diffusion delay
+--
+-- This is the number of slots a 'CS.MsgRollForward' should arrive after the
+-- forge slot of the header it is carrying.
+--
+-- It may depend on the @(sender, recipient)@, the current slot, and header.
+newtype CalcMessageDelay blk = CalcMessageDelay
```

Comment:
Something to consider for later, perhaps: do you envision doing any complicated calculation here, or would a mere tabulation (e.g. `[(CoreNodeId, CoreNodeId, SlotNo, NumSlots)]`) suffice? If so, that might be easier to `Show`, interpret in test failures, and, critically, shrink.

Reply:
I copied this over from the branch I've been developing semi-synchrony on for Praos. On that branch, this computation is driven by a random seed. That's where this shape originated. I'm hoping to develop that next and share this infra.

I don't have much intuition for the "size" of a set of delays. The main obstacle seems to be that the actual set of messages being delayed is unpredictable (especially for Praos). And so I'm punting on shrinking for now.

To your specific suggestion: `[(CoreNodeId, CoreNodeId, SlotNo, NumSlots)]` seems to assume that there is only one delay per directed edge per slot. But two nodes sometimes exchange multiple messages per slot. I'm unsure if it matters that those message potentially have different delays, but I'm currently favoring that they can so-as to make the fewest assumptions.